Tests/SwiftLintFrameworkTests/BaselineTests.swift

```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    func testUnchangedViolations() throws {
+        try testBlock { XCTAssertEqual(baseline.filter(violations), []) }
+    }
+
+    func testShiftedViolations() throws {
+        try testBlock {
+            XCTAssertEqual(baseline.filter(try violations.lineShifted(by: 2, path: sourceFilePath)), [])
+        }
+    }
+
+    func testNewViolation() throws {
+        try testViolationDetection(
+            violations: violations,
+            newViolationRuleDescription: EmptyCollectionLiteralRule.description,
+            insertionIndex: 2
+        )
+    }
+
+    func testViolationsWithNoFile() throws {
+        try testViolationDetection(
+            violations: violations.map { $0.with(location: Location(file: nil)) },
+            lineShift: 0,
+            newViolationRuleDescription: ArrayInitRule.description,
+            insertionIndex: 2
+        )
+    }
+
+    func testViolationDetection() throws {
+        let violations = [
+            ArrayInitRule.description,
+            BlockBasedKVORule.description,
+            ArrayInitRule.description,
+            ClosingBraceRule.description,
+            ClosingBraceRule.description,
+            ClosingBraceRule.description,
+            BlockBasedKVORule.description,
+            DirectReturnRule.description,
+            ArrayInitRule.description,
+            ClosingBraceRule.description
+        ].violations
+
+        let ruleDescriptions = [
+            ArrayInitRule.description,
+            BlockBasedKVORule.description,
+            ClosingBraceRule.description,
+            DirectReturnRule.description
+        ]
+
+        for ruleDescription in ruleDescriptions {
+            for insertionIndex in 0..<violations.count {
+                try testViolationDetection(
+                    violations: violations,
+                    newViolationRuleDescription: ruleDescription,
+                    insertionIndex: insertionIndex
+                )
+            }
+        }
+    }
+
+    private func testViolationDetection(
+        violations: [StyleViolation],
+        lineShift: Int = 1,
+        newViolationRuleDescription: RuleDescription,
+        insertionIndex: Int
+    ) throws {
+        try testBlock {
+            let baseline = Baseline(violations: violations)
+            var newViolations = lineShift != 0
+                ? try violations.lineShifted(by: lineShift, path: sourceFilePath)
+                : violations
+            let line = ((insertionIndex + 1) * 2) - 1 + lineShift
+            let violation = StyleViolation(
+                ruleDescription: newViolationRuleDescription,
+                location: Location(file: sourceFilePath, line: line, character: 1)
+            )
+            newViolations.insert(violation, at: insertionIndex)
+            XCTAssertEqual(baseline.filter(newViolations), [violation])
+        }
+    }
+
+    private func testBlock(_ block: () throws -> Void) throws {
```

Comment:
I wonder if this can also be achieved by implementing appropriate `setUp` and `tearDown` methods in a test class.

Reply:
So I've moved the directory changing methods into `setUpWithError` and `tearDownWithError`.

For the test file, which we need to create on disk before each test, setUp would almost have worked, but in `testViolationDetection` we need to perform that setup before testing each rule in each position.